edgelet/iotedged/src/lib.rs

```diff
@@ -1533,25 +1529,22 @@ where
     };
 
     let edge_rt_with_mgmt_signal = edge_rt.select2(mgmt_stop_and_reprovision_signaled).then(
-        |res: Result<
-            Either<((), _), (Option<Error>, _)>,
-            Either<(Error, _), (Option<Error>, _)>,
-        >| {
-            // A -> EdgeRt Future
-            // B -> Mgmt Stop and Reprovision Signal Future
-            match res {
-                Ok(Either::A((_x, _y))) => {
-                    Ok((StartApiReturnStatus::Shutdown, false)).into_future()
-                }
-                Ok(Either::B((_x, _y))) => {
-                    debug!("Shutdown with device reprovisioning.");
-                    Ok((StartApiReturnStatus::Shutdown, true)).into_future()
-                }
-                Err(Either::A((err, _y))) => Err(err).into_future(),
-                Err(Either::B((err, _y))) => {
-                    debug!("The mgmt shutdown and reprovision signal failed.");
-                    Err(err.unwrap()).into_future()
-                }
+        |res| match res {
+            Ok(Either::A((_edge_rt_ok, _mgmt_stop_and_reprovision_signaled_future))) => {
+                info!("Edge runtime will stop because of the shutdown signal.");
+                future::ok((StartApiReturnStatus::Shutdown, false))
```

Comment:
The return value for this gigantic function seem like they could be represented as a tri-state enum than a (enum, bool) combination, where the bool really matters to just one enum.